File: EnglishSpeakingCoach/EnglishSpeakingCoach.Infrastructure/Services/RecordingService.cs
using AutoMapper;
using EnglishSpeakingCoach.Application.DTOs;
using EnglishSpeakingCoach.Application.Services;
using EnglishSpeakingCoach.Domain.Entities;
using EnglishSpeakingCoach.Domain.Interfaces;
using Microsoft.EntityFrameworkCore;
using EnglishSpeakingCoach.Infrastructure.Data;

namespace EnglishSpeakingCoach.Infrastructure.Services;

public class RecordingService : IRecordingService
{
    private readonly IUnitOfWork _unitOfWork;
    private readonly IMapper _mapper;
    private readonly IMLService _mlService;
    private readonly ApplicationDbContext _context;

    public RecordingService(
        IUnitOfWork unitOfWork,
        IMapper mapper,
        IMLService mlService,
        ApplicationDbContext context)
    {
        _unitOfWork = unitOfWork;
        _mapper = mapper;
        _mlService = mlService;
        _context = context;
    }

    public async Task<RecordingDto> UploadRecordingAsync(UploadRecordingRequest request)
    {
        var recording = _mapper.Map<Recording>(request);
        recording.CreatedAt = DateTime.UtcNow;
        recording.ProcessingStatus = "Pending";

        await _unitOfWork.Recordings.AddAsync(recording);
        await _unitOfWork.SaveChangesAsync();

        // Trigger async processing (in real app, this would be a background job)
        _ = Task.Run(async () => await ProcessRecordingAsync(recording.Id));

        return _mapper.Map<RecordingDto>(recording);
    }

    public async Task<RecordingDto?> GetRecordingByIdAsync(int id)
    {
        var recording = await _unitOfWork.Recordings.GetByIdAsync(id);
        return recording == null ? null : _mapper.Map<RecordingDto>(recording);
    }

    public async Task<IEnumerable<RecordingDto>> GetSessionRecordingsAsync(int sessionId)
    {
        var recordings = await _unitOfWork.Recordings.FindAsync(r => r.SessionId == sessionId);
        return _mapper.Map<IEnumerable<RecordingDto>>(recordings.OrderBy(r => r.CreatedAt));
    }

    public async Task<RecordingAnalysisDto?> GetRecordingAnalysisAsync(int recordingId)
    {
        var recording = await _context.Recordings
            .Include(r => r.Feedbacks)
            .FirstOrDefaultAsync(r => r.Id == recordingId);

        if (recording == null)
        {
            return null;
        }

        var feedbacks = _mapper.Map<List<FeedbackDto>>(recording.Feedbacks);

        return new RecordingAnalysisDto(
            RecordingId: recording.Id,
            Transcript: recording.Transcript,
            PronunciationScore: recording.PronunciationScore,
            FluencyScore: recording.FluencyScore,
            AccuracyScore: recording.AccuracyScore,
            Feedbacks: feedbacks
        );
    }

    public async Task ProcessRecordingAsync(int recordingId)
    {
        var recording = await _unitOfWork.Recordings.GetByIdAsync(recordingId);
        if (recording == null)
        {
            return;
        }

        try
        {
            recording.ProcessingStatus = "Processing";
            await _unitOfWork.Recordings.UpdateAsync(recording);
            await _unitOfWork.SaveChangesAsync();

            // Call ML service for analysis
            var analysis = await _mlService.AnalyzeRecordingAsync(recording.AudioUrl);

            // Update recording with analysis results
            recording.Transcript = analysis.Transcript;
            recording.PronunciationScore = analysis.PronunciationScore;
            recording.FluencyScore = analysis.FluencyScore;
            recording.AccuracyScore = analysis.AccuracyScore;
            recording.ProcessingStatus = "Completed";
            recording.ProcessedAt = DateTime.UtcNow;

            await _unitOfWork.Recordings.UpdateAsync(recording);

            // Save feedback items
            foreach (var feedbackItem in analysis.FeedbackItems)
            {
                var feedback = new Feedback
                {
                    RecordingId = recordingId,
                    FeedbackType = feedbackItem.Type,
                    Content = feedbackItem.Content,
                    DetailedAnalysis = feedbackItem.DetailedAnalysis,
                    Severity = feedbackItem.Severity,
                    WordPosition = feedbackItem.WordPosition,
                    Suggestion = feedbackItem.Suggestion,
                    CreatedAt = DateTime.UtcNow
                };
                await _unitOfWork.Feedbacks.AddAsync(feedback);
            }

            await _unitOfWork.SaveChangesAsync();
        }
        catch
        {
            recording.ProcessingStatus = "Failed";
            await _unitOfWork.Recordings.UpdateAsync(recording);
            await _unitOfWork.SaveChangesAsync();
            throw;
        }
    }
}

File: english-speaking-coach-app/src/app/core/services/recording.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface Recording {
  id: number;
  sessionId: number;
  audioUrl: string;
  transcript?: string;
  pronunciationScore: number;
  fluencyScore: number;
  accuracyScore: number;
  overallScore: number;
  duration: number;
  createdAt: string;
}

export interface CreateRecordingRequest {
  sessionId: number;
  audioUrl: string;
  duration: number;
}

export interface RecordingFeedback {
  recordingId: number;
  transcript: string;
  pronunciationScore: number;
  fluencyScore: number;
  accuracyScore: number;
  overallScore: number;
  feedbacks: Feedback[];
}

export interface Feedback {
  id: number;
  recordingId: number;
  feedbackType: string;
  content: string;
  detailedAnalysis?: string;
  severity?: number;
  wordPosition?: number;
  suggestion?: string;
  createdAt: string;
}

@Injectable({
  providedIn: 'root'
})
export class RecordingService {
  constructor(private http: HttpClient) {}

  createRecording(recording: CreateRecordingRequest): Observable<Recording> {
    return this.http.post<Recording>(`${environment.apiUrl}/recordings`, recording);
  }

  getRecording(id: number): Observable<Recording> {
    return this.http.get<Recording>(`${environment.apiUrl}/recordings/${id}`);
  }

  getUserRecordings(userId: number): Observable<Recording[]> {
    return this.http.get<Recording[]>(`${environment.apiUrl}/recordings/user/${userId}`);
  }

  analyzeRecording(recordingId: number): Observable<RecordingFeedback> {
    return this.http.post<RecordingFeedback>(
      `${environment.apiUrl}/recordings/${recordingId}/analyze`, 
      {}
    );
  }

  getRecordingFeedback(recordingId: number): Observable<Feedback[]> {
    return this.http.get<Feedback[]>(`${environment.apiUrl}/recordings/${recordingId}/feedback`);
  }
}
